File: Dynamic Programming/72. Edit Distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        dp = [[float("inf") for i in range(len(word2) + 1)] for j in range(len(word1) + 1)]
        
        dp[0][0] = 0 
        
        # row 
        for i in range(len(word1) + 1):
            dp[i][0] = i

        for j in range(len(word2) + 1):
            dp[0][j] = j 

        for i in range(1,len(word1) + 1):
            for j in range(1,len(word2) + 1):
                toAdd = 0
                if word2[j - 1] != word1[i - 1]:
                    toAdd += 1 
                dp[i][j] = min(dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] + toAdd)
                
                
                
                
        return dp[-1][-1]

File: strings/72. Edit distance.cpp
#include <algorithm>
class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.length();
        int n = word2.length();

        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));

        for (int i = 1; i <= n; ++i) {
            dp[0][i] = 1 + dp[0][i - 1];
        }

        word1 = " " + word1;
        word2 = " " + word2;

        for (int j = 1; j <= m; ++j) {
            dp[j][0] = 1 + dp[j - 1][0];
        }

        for (int i = 1; i <= m; ++i) {
            for (int j = 1; j <= n; ++j) {
                if (word1[i] == word2[j]) {
                    dp[i][j] = min({
                        dp[i - 1][j - 1],
                        dp[i - 1][j] + 1,
                        dp[i][j - 1] + 1
                    });
                } else {
                    dp[i][j] = min({
                        dp[i - 1][j - 1],
                        dp[i - 1][j],
                        dp[i][j - 1]
                    }) + 1;
                }
            }
        }


        return dp[m][n];
    }
};